plugin/forward/forward.go

```diff
@@ -68,6 +72,14 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 		return plugin.NextOrFailure(f.Name(), f.Next, ctx, w, r)
 	}
 
+	if f.maxQueryCount > 0 {
+		count := atomic.AddInt64(&(f.queryCount), 1)
+		defer atomic.AddInt64(&(f.queryCount), -1)
+		if count > f.maxQueryCount {
+			return dns.RcodeServerFailure, errors.New("inflight forward queries exceeded maximum")
```

Comment:
Would refuse be better here?

Reply:
I'm not sure.  I think either could work. I initially thought REFUSED, because this is a known/expected failure (soft failure). ... but then changed it to SERVFAIL in a later commit after reading some discussions on the difference between the two. 

There isn't a strict difference: It's fuzzy...
REFUSE is supposed to indicate refusal due to policy. I read that as a function of the client or queried name.  But thats vague, and you could describe a query rate type limit like this as a "policy".
SERVFAIL is supposed to indicate something broken.